File: Orbit_ADCS/Matlab-ADCS/ADCS Julia/Determination/OLAE.jl
include("../types.jl")

function OLAE(vkb::Matrix{Float64}, vkn::Matrix{Float64}, weight::Vector{Float64})
    d::Vector{Float64} = vec(vkb - vkn)
    N::Int64 = length(d)
    S::Matrix{Float64} = zeros(N,3)
    for i = 1:Int(N/3)
        S[i:i+2,:] = cross2mat(vkb[:,i] + vkn[:,i])
    end
    W::Matrix{Float64} = Matrix(I,N,N)
    qcr = CRP(inv(S'*W*S)*S'*W*d, "B", "N")

    return CRP2dcm(qcr)
end

File: Orbit_ADCS/Matlab-ADCS/Determination/OLAE.m
function var = OLAE(vkb, vkn, weight)
    TP = types;
    CRP = CRPRot;
    d0 = vkb - vkn;
    d = reshape(d0,[numel(d0),1]);
    N = length(d);
    S = zeros(N,3);
    for i = 1:(N/3)
        S(i:i+2,:) = skewsym(vkb(:,i) + vkn(:,i));
        for j = 3*i - [2 1 0]
            W(j,j) = weight(i);
        end
    end
    SWS = (S'*W*S);
    if 1/cond(SWS) < 1e-4
        warning("Matrix is close to singular or badly scaled. Results may be inaccurate. Changing to QUEST method")
        var = QUEST(vkb, vkn, weight);
        return
    end
    qcr = TP.CRP(SWS\(S'*W*d), "B", "N");

    var = CRP.CRP2dcm(qcr);
end